Index: linux-xilinx-v2024.1/drivers/net/phy/intel-xway.c
===================================================================
--- linux-xilinx-v2024.1.orig/drivers/net/phy/intel-xway.c
+++ linux-xilinx-v2024.1/drivers/net/phy/intel-xway.c
@@ -252,6 +252,10 @@ static int xway_gphy_config_init(struct
 	if (err)
 		return err;
 
+	err = phy_write(phydev, 0x17, 0xb500);
+	if (err)
+		return err;
+
 	phy_write_mmd(phydev, MDIO_MMD_VEND2, XWAY_MMD_LEDCH,
 		      XWAY_MMD_LEDCH_NACS_NONE |
 		      XWAY_MMD_LEDCH_SBF_F02HZ |
