Index: u-boot-2016.01-rc3/arch/arm/mach-zynq/Kconfig
===================================================================
--- u-boot-2016.01-rc3.orig/arch/arm/mach-zynq/Kconfig
+++ u-boot-2016.01-rc3/arch/arm/mach-zynq/Kconfig
@@ -21,6 +21,9 @@ config TARGET_ZYNQ_MICROZED
 config TARGET_ZYNQ_PICOZED
 	bool "Zynq PicoZed"
 
+config TARGET_ZYNQ_RED_PITAYA
+	bool "Red Pitaya Board"
+
 config TARGET_ZYNQ_ZC702
 	bool "Zynq ZC702 Board"
 
@@ -49,6 +52,7 @@ config SYS_CONFIG_NAME
 	default "zynq_zed" if TARGET_ZYNQ_ZED
 	default "zynq_microzed" if TARGET_ZYNQ_MICROZED
 	default "zynq_picozed" if TARGET_ZYNQ_PICOZED
+	default "zynq_red_pitaya" if TARGET_ZYNQ_RED_PITAYA
 	default "zynq_zc70x" if TARGET_ZYNQ_ZC702 || TARGET_ZYNQ_ZC706
 	default "zynq_zc770" if TARGET_ZYNQ_ZC770
 	default "zynq_zybo" if TARGET_ZYNQ_ZYBO
Index: u-boot-2016.01-rc3/configs/zynq_red_pitaya_defconfig
===================================================================
--- /dev/null
+++ u-boot-2016.01-rc3/configs/zynq_red_pitaya_defconfig
@@ -0,0 +1,20 @@
+CONFIG_ARM=y
+CONFIG_ARCH_ZYNQ=y
+CONFIG_TARGET_ZYNQ_RED_PITAYA=y
+CONFIG_DEFAULT_DEVICE_TREE="zynq-red-pitaya"
+CONFIG_SPL=y
+CONFIG_FIT=y
+CONFIG_FIT_VERBOSE=y
+CONFIG_FIT_SIGNATURE=y
+# CONFIG_CMD_IMLS is not set
+# CONFIG_CMD_FLASH is not set
+CONFIG_CMD_GPIO=y
+# CONFIG_CMD_SETEXPR is not set
+CONFIG_NET_RANDOM_ETHADDR=y
+CONFIG_OF_SEPARATE=y
+CONFIG_ZYNQ_GEM=y
+CONFIG_ZYNQ_QSPI=y
+CONFIG_DEBUG_UART=y
+CONFIG_DEBUG_UART_ZYNQ=y
+CONFIG_DEBUG_UART_BASE=0xe0000000
+CONFIG_DEBUG_UART_CLOCK=50000000
