Index: u-boot-xlnx-xlnx_rebase_v2026.01_2026.1/env/eeprom.c
===================================================================
--- u-boot-xlnx-xlnx_rebase_v2026.01_2026.1.orig/env/eeprom.c
+++ u-boot-xlnx-xlnx_rebase_v2026.01_2026.1/env/eeprom.c
@@ -18,6 +18,10 @@
 #include <errno.h>
 #include <linux/compiler.h>	/* for BUG_ON */

+#ifndef CONFIG_SYS_I2C_EEPROM_BUS
+#define CONFIG_SYS_I2C_EEPROM_BUS -1
+#endif
+
 DECLARE_GLOBAL_DATA_PTR;

 static int env_eeprom_load(void)
@@ -31,7 +35,7 @@
 	uchar rdbuf[64], flags[2];
 	int i, crc_ok[2] = {0, 0};

-	eeprom_init(-1);	/* prepare for EEPROM read/write */
+	eeprom_init(CONFIG_SYS_I2C_EEPROM_BUS); /* prepare for EEPROM read/write */

 	off_env[0] = CONFIG_ENV_OFFSET;
 	off_env[1] = CONFIG_ENV_OFFSET_REDUND;
@@ -91,7 +95,7 @@
 	ulong crc, len, new;
 	uchar rdbuf[64];

-	eeprom_init(-1);	/* prepare for EEPROM read/write */
+	eeprom_init(CONFIG_SYS_I2C_EEPROM_BUS); /* prepare for EEPROM read/write */

 	/* read old CRC */
 	eeprom_read(CONFIG_SYS_I2C_EEPROM_ADDR,
