Index: u-boot-2017.01/include/configs/zynq-common.h
===================================================================
--- u-boot-2017.01.orig/include/configs/zynq-common.h
+++ u-boot-2017.01/include/configs/zynq-common.h
@@ -208,6 +208,7 @@
 			"run setbootenv \0" \
 	"usb_loadbootenv=set bootenv_dev usb && usb start && run setbootenv \0" \
 	"preboot=if test $modeboot = sdboot; then " \
+			"mmc rescan;" \
 			"run sd_loadbootenv; " \
 			"echo Checking if uenvcmd is set ...; " \
 			"if test -n $uenvcmd; then " \
