Index: linux-xilinx-v2019.1/scripts/dtc/dtc-lexer.l
===================================================================
--- linux-xilinx-v2019.1.orig/scripts/dtc/dtc-lexer.l
+++ linux-xilinx-v2019.1/scripts/dtc/dtc-lexer.l
@@ -38,7 +38,6 @@ LINECOMMENT	"//".*\n
 #include "srcpos.h"
 #include "dtc-parser.tab.h"
 
-YYLTYPE yylloc;
 extern bool treesource_error;
 
 /* CAUTION: this will stop working if we ever use yyless() or yyunput() */
